ServiceWorkerContainer::Promise

Coder206 opened the pull request:
<!-- Please describe your changes on the following line: -->
## The purpose of the code changes is to enable the use of promises in the Service Worker container. I also modified the Service Worker test in order to support the promises.

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #13409  (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13419)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/package_commands.py
 * @kichjang: components/script/dom/serviceworkercontainer.rs, components/script/dom/webidls/ServiceWorkerContainer.webidl

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

Coder206 commented on components/script/dom/serviceworkercontainer.rs:
```diff
@@ -57,6 +66,26 @@ impl ServiceWorkerContainerMethods for ServiceWorkerContainer {
+        
+        // New Step 1
+        
+        let p = dom::promise::Promise
```
Is this right?

jdm replied:
I recommend looking for uses of `Promise::new` in the codebase to base your changes off of.

Coder206 commented on components/script/dom/webidls/ServiceWorkerContainer.webidl:
```diff
@@ -6,21 +6,21 @@
 interface ServiceWorkerContainer : EventTarget {
   [Unforgeable] readonly attribute ServiceWorker? controller;
-  //[SameObject] readonly attribute Promise<ServiceWorkerRegistration> ready;
+  [SameObject] readonly attribute Promise<ServiceWorkerRegistration> ready;
```
Just preparing for eventual functionnality

jdm replied:
This should be replacing the `register` underneath it.

Coder206 commented:
It's the first time I do one of these, I am new to the process and I am having trouble understanding how to translate the pseudo code from [https://w3c.github.io/ServiceWorker/#service-worker-container-register-method](Service Worker Nightly Container:Register) into Rust for Servo

Coder206 commented:
I think this PR is for @jdm

jdm commented on components/script/dom/serviceworkercontainer.rs:
```diff
@@ -56,7 +65,27 @@ impl ServiceWorkerContainerMethods for ServiceWorkerContainer {
+        Let scopeURL be null.
+        If options.scope is present, set scopeURL to the result of parsing options.scope with the context object’s relevant settings object’s API base URL.
+        Invoke Start Register with scopeURL, scriptURL, p, client, client’s creation URL and options.type.
+        Return p.    */ 
```
As mentioned in #13409, I don't think there's any need to focus on the specification. It's enough to know that it says to create a promise and return it, and either resolve it with a successful service worker or else reject it with an error (instead of returning error values directly).

jdm commented on components/script/dom/webidls/ServiceWorkerContainer.webidl:
```diff
@@ -6,21 +6,21 @@
-  //[NewObject] /*Promise<any>*/ any getRegistration(optional USVString clientURL = "");
-  //[NewObject] /* Promise */<sequence<ServiceWorkerRegistration>> getRegistrations();
+  [NewObject] Promise<any> any getRegistration(optional USVString clientURL = "");
+  [NewObject] Promise <sequence<ServiceWorkerRegistration>> getRegistrations();
```
We shouldn't be modifying any other methods or properties in this file for the purposes of these changes.

jdm reviewed (changes requested):
You will definitely need to be able to build Servo locally to be able to make these changes effectively.

Coder206 commented on components/script/dom/webidls/ServiceWorkerContainer.webidl:
```diff
@@ -6,9 +6,7 @@
-  //[SameObject] readonly attribute Promise<ServiceWorkerRegistration> ready;
-
-  [NewObject, Throws] ServiceWorkerRegistration register(USVString scriptURL, optional RegistrationOptions options);
+  [SameObject] readonly attribute Promise<ServiceWorkerRegistration> ready;
```
@jdm When running `./mach build -d`, I am seeing this error:

`WebIDL.WebIDLError: error: Unresolved type '<unresolved scope>::Promise'., /home/user/servo/components/script/dom/webidls/ServiceWorkerContainer.webidl line 9:34
  [SameObject] readonly attribute Promise<ServiceWorkerRegistration> ready;`

jdm replied:
Your branch is based off of a master commit from Sept 6. You need to be using 2b1a39c2ae14b78e437551d05f0e691a13e5d183 or more recent instead.

creativcoder replied:
@Coder206  As an added note, We shouldn't be needing to make changes to the `ready` method for the relevant issue. You just have to make changes to the `register` api declaration from:
```
[NewObject, Throws] ServiceWorkerRegistration register(USVString scriptURL, optional RegistrationOptions options);
```

 to 

```
[NewObject] Promise<ServiceWorkerRegistration> register(USVString scriptURL, optional RegistrationOptions options);
```
After that you need to either use a combination of [new_with_js_promise](https://github.com/servo/servo/blob/d00639c55f9a342765483d347a3c29d4647f7411/components/script/dom/promise.rs#L91) and [resolve_native](https://github.com/servo/servo/blob/d00639c55f9a342765483d347a3c29d4647f7411/components/script/dom/promise.rs#L139) methods to make the register method return a resolved promise. Correct me if am wrong @jdm

Rebase your current branch to the master branch before making these changes as mentioned by jdm

Coder206 replied:
@creativcoder I am not sure how to get started with `resolve_native` and `reject_error`. How do I implement that in my code? (does it go in the [return](https://github.com/servo/servo/pull/13419/files/b4b7fa5003454846f15ce0cadecbc726697b06cf#diff-2c27b3453cfacbe80bdac4094a0ad0c9R68))

creativcoder replied:
We don't need to implement those methods, they are already implemented. In the `Register` method we just need to call `reject_error` on places where we return an `Err` value and `resolve_native` at places where we return an `Ok` value on the promise object.

You can have a look at [my branch](https://github.com/servo/servo/compare/master...creativcoder:swcontainer_promise?expand=1) here, for how that is being done.

Coder206 commented:
@jdm So I need to rebase my code to the newest commit?

jdm commented:
@Coder206 That would be easiest, yes.

Coder206 commented:
Looks like an extra commit was added here. I am going to remove it later today.

jdm commented:
The test file mentioned in the original issue still needs modifying, too.

Coder206 commented:
@jdm Thanks for the reminder

KiChjang commented on components/script/dom/serviceworkercontainer.rs:
```diff
@@ -53,53 +55,75 @@ impl ServiceWorkerContainerMethods for ServiceWorkerContainer {
         if scope.path().to_ascii_lowercase().contains("%2f") ||
         scope.path().to_ascii_lowercase().contains("%5c") {
-            return Err(Error::Type("Scope URL contains forbidden characters".to_owned()));
+            return {
```
What's the purpose of this `return` keyword here?

jdm replied:
This should still be addressed.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13474) made this pull request unmergeable. Please resolve the merge conflicts.

Coder206 committed:
Add Promises to Service Worker Container

The purpose of the code changes is to enable the use of promises in the Service Worker container. I also modified the Service Worker test in order to support the promises.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -1,35 +1,43 @@
-    console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
-    document.getElementById('iframe_sw').src = 'demo_iframe.html';
-    post_message('/demo_iframe');
+        /* global navigator */
```
I'm not sure what this comment means.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -1,35 +1,43 @@
+            console.log("From client worker registered: ", navigator.serviceWorker.controller);
+            console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
+        }).catch(function(error) {
+            console.log("Error in loading: " + navigator.serviceWorker.controller.scriptURL);
```
Same as previous.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -40,17 +48,23 @@
+            console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
+            post_message('/dashboard');
+        }).catch(function(error) {
+            console.log("Error in loading: " + navigator.serviceWorker.controller.scriptURL);
```
Same.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -40,17 +48,23 @@
+            console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
+            post_message('/profile');
+        }).catch(function(error) {
+            console.log("Error in loading: " + navigator.serviceWorker.controller.scriptURL);
```
Same.

jdm reviewed (changes requested):
This looks good with a few point addressed.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13596) made this pull request unmergeable. Please resolve the merge conflicts.

Coder206 commented on tests/html/service-worker/index.html:
```diff
@@ -1,35 +1,44 @@
+            document.getElementById('iframe_sw').src = 'demo_iframe.html';
+            post_message('/demo_iframe'); 
+        }).catch(function(error) {
+            console.log("Error in loading `iframe_sw.js` --> " + error);
```
@jdm I keep getting this error when running the test: `TypeError: Only secure origins allowed` on each of the equivalents below.

jdm replied:
I believe you're running the test from a file:// URL, rather than via an http:// URL. If you set up an HTTP server (such as running `python -m SimpleHTTPServer`) and visit `http://localhost:8000/index.html` then it should work as expected.

Coder206 replied:
@jdm Ok that makes sense.

Coder206 commented on tests/html/service-worker/index.html:
```diff
@@ -40,17 +49,25 @@
+            console.log("Registered script source: " + "sw_profile.js");
+            post_message('/profile');
+        }).catch(function(error) {
+            console.log("Error in loading `sw_profile.js` --> " + error);
```
Equivalent

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -21,22 +21,23 @@
     <script>
-        /* global navigator */
-        navigator.serviceWorker.register('iframe_sw.js', { 'scope': '/demo_iframe.html' }).then(function(registration) {
+        navigator.serviceWorker.register('/tests/html/service-worker/iframe_sw.js', { 'scope': '/tests/html/service-worker/'})
```
This change should be reverted - it seems to be relying on having an HTTP server running from the root Servo directory, but I think we should assume a server running with tests/html/service-worker/ as the root instead.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -21,22 +21,23 @@
+        .then(function(registration) {
             console.log("From client worker registered: ", navigator.serviceWorker.controller);
-            console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
+            console.log("Registered script source: " + "iframe_sw.js");
```
These can be combined into a single string.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -21,22 +21,23 @@
+        .then(function(registration) {
             console.log("From client worker registered: ", navigator.serviceWorker.controller);
-            console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
+            console.log("Registered script source: " + "sw.js");
```
Combine into one string.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -48,22 +49,24 @@
+        .then(function(registration) {
             console.log("From client worker registered: ", navigator.serviceWorker.controller);
-            console.log("Registered script source: "+ navigator.serviceWorker.controller.scriptURL);
+            console.log("Registered script source: " + "sw_profile.js");
```
Combine into one string.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -1,35 +1,42 @@
+        <button onclick="register_zero()">Register for Root (/)</button>
+        <button onclick="register_one()">Register for Dashboard (/dashboard)</button>
+        <button onclick="register_two()">Register for Profile (/profile)</button>
+        <button onclick="post_messaget()">Msg to Root service worker</button>
```
This looks like an accidental change.

jdm commented on tests/html/service-worker/index.html:
```diff
@@ -1,35 +1,42 @@
-    post_message('/demo_iframe');
+        navigator.serviceWorker.register('iframe_sw.js', { 'scope': '/demo_iframe.html'})
+        .then(function(registration) {
+            console.log("From client worker registered: ", navigator.serviceWorker.controller + "\nRegistered script source: " + "iframe_sw.js");
```
When I said combine into one string, I didn't mean combining two console.log invocations into one. I meant the pattern of `console.log("a literal string" + "another literal string")` can become `console.log("a literal string another literal string")` instead.

jdm reviewed (changes requested):
One last one that was overlooked! Otherwise this looks fine :)

jdm commented on components/script/dom/serviceworkercontainer.rs:
```diff
@@ -4,11 +4,12 @@
 use dom::bindings::js::{JS, MutNullableHeap, Root};
 use dom::bindings::reflector::{Reflectable, reflect_dom_object};
 use dom::bindings::str::USVString;
+use dom::bindings::js::RootedReference;
```
This will fail `./mach test-tidy`.

jdm commented:
@bors-servo: r+
Thanks!

bors-servo commented:
:pushpin: Commit 326aed6 has been approved by `jdm`

<!-- @bors-servo r=jdm 326aed613ac85cae7d15c8bfd8506bccda0995ce -->

bors-servo commented:
:hourglass: Testing commit 326aed6 with merge 702aa31...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/249)

jdm commented:
Oh look, we have some automated tests too! We'll need to update those to use promises now.

jdm commented:
I filed #13776 for the hr-time failure.

Coder206 commented on components/script/dom/serviceworkercontainer.rs:
```diff
@@ -53,55 +56,74 @@ impl ServiceWorkerContainerMethods for ServiceWorkerContainer {
                 script_url: USVString,
-                options: &RegistrationOptions) -> Fallible<Root<ServiceWorkerRegistration>> {
+                options: &RegistrationOptions) -> Rc<Promise> {
+        let promise = Promise::new(self.global().r());
```
@jdm I was chatting with @KiChjang on IRC and I was told not to use `.r()` anymore in my code. Would you want me to carry out the changes towards a non-`r()` code before you merge?

jdm replied:
Sure, you can remove the two uses of `.r()` in these changes.

jdm commented:
Looks like `./mach test-wpt tests/wpt/mozilla/tests/mozilla/service-workers/service-worker-registration.html` will still fail.

Coder206 commented:
@jdm Is there anything I can do to help out? (Is it hard to update the tests)

jdm commented:
@Coder206 It is your responsibility to update it, yes :) You'll need to perform the following transformations in each test that calls `register_sw`:
* use `promise_test` instead of `test`
* move the `assert_foo` checks into `then()` callbacks
As an example, if the existing code looks like:
```js
test(function() {
  var sw_reg = register_sw('sw.js');
  assert_class_string(sw_reg, "ServiceWorkerRegistration");
}, 'some message');
```
We need to change it like so:
```js
promise_test(function() {
    return register_sw('sw.js').then(function(sw_reg) {
        assert_class_string(sw_reg, "ServiceWorkerRegistration");
    });
}, 'some message');
```

jdm commented:
Additionally, the tests that use assert_throws should instead use `promise_rejects` as described in the testharness.js [documentation](http://testthewebforward.org/docs/testharness-library.html) (search for `project_rejects`).

Coder206 commented:
@jdm OK sounds good, on it!

Coder206 commented:
@jdm I made some changes to the WPT test `service-worker.registration.html` based on your comments. I was receiving an error when running the test: `Unsupported test type wdspec for product servo` but I tried another test https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/activation.html and received the same outcome. So I am thinking that my setup is incompatible with the mozilla wpt tests. Do this make sense?

jdm commented on tests/wpt/mozilla/tests/mozilla/service-workers/service-worker-registration.html:
```diff
@@ -10,43 +10,47 @@
 }
 
-test(function (){
+promise_test(function (){
```
This can stay `test`, since it doesn't use promises.

jdm reviewed (changes requested):
These changes look great! Only one small revert required before we can merge this.

jdm commented:
The `Unsupported test type wdspec for product servo` should not be present on a recent revision based on master. It also is just an informational message; it doesn't affect the results of the tests at all.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit a02ccef has been approved by `jdm`

<!-- @bors-servo r=jdm a02ccef697103cc4eb0165eb35760781a5048fb5 -->

bors-servo commented:
:hourglass: Testing commit a02ccef with merge 6a6bfcf...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/450)

KiChjang commented:
```
Tests with unexpected results:
  ▶ Unexpected subtest result in /_mozilla/mozilla/service-workers/service-worker-registration.html:
  │ FAIL [expected PASS] Test: Active Service Worker ScriptURL property
  └   → promise_test: Unhandled rejection with value: object "ReferenceError: sw_reg is not defined"

  ▶ Unexpected subtest result in /_mozilla/mozilla/service-workers/service-worker-registration.html:
  │ FAIL [expected PASS] Test: Throws Error when Invalid Url Path
  │   → promise.then is not a function
  │ 
  │ promise_rejects@http://web-platform.test:8000/resources/testharness.js:547:16
  │ @http://web-platform.test:8000/_mozilla/mozilla/service-workers/service-worker-registration.html:34:10
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:530:36

  ▶ Unexpected subtest result in /_mozilla/mozilla/service-workers/service-worker-registration.html:
  │ FAIL [expected PASS] Test: Throws Error when Invalid Scope
  │   → promise.then is not a function
  │ 
  │ promise_rejects@http://web-platform.test:8000/resources/testharness.js:547:16
  │ @http://web-platform.test:8000/_mozilla/mozilla/service-workers/service-worker-registration.html:40:10
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:530:36
```

Coder206 commented:
I think I saw my mistake on the first error. I left out the `sw_reg` from the function parameters. I am not sure what the other two are about, planning on looking into this further.

jdm commented:
@Coder206 Did you run the test after making the changes? My assumption when people make changes to test code is that the tests pass before they push their changes.

Coder206 commented:
@jdm Yes, that's when I received the error from above. I haven't rebased though with the new changes. The next push will have the tests verified with the new master.

Coder206 commented:
Sorry for the inconvenience.

Coder206 commented:
I am unable to get the error messages that @KiChjang has shared with me. Here is the output of my attempt to test my changes. Please note, this is the result despite I only corrected the first error mentioned above: 

![page-shot-2016-10-30 novnc](https://cloud.githubusercontent.com/assets/7758155/19837258/673085e8-9e8c-11e6-89a8-3670e4415b82.png)

Errors I am unsure what they mean: 

```
▶ Unexpected subtest result in /_mozilla/mozilla/service-workers/service-worker-registration.html:
  │ FAIL [expected PASS] Test: Throws Error when Invalid Url Path
  │   → promise.then is not a function
  │ 
  │ promise_rejects@http://web-platform.test:8000/resources/testharness.js:547:16
  │ @http://web-platform.test:8000/_mozilla/mozilla/service-workers/service-worker-registration.html:34:10
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:530:36

  ▶ Unexpected subtest result in /_mozilla/mozilla/service-workers/service-worker-registration.html:
  │ FAIL [expected PASS] Test: Throws Error when Invalid Scope
  │   → promise.then is not a function
  │ 
  │ promise_rejects@http://web-platform.test:8000/resources/testharness.js:547:16
  │ @http://web-platform.test:8000/_mozilla/mozilla/service-workers/service-worker-registration.html:40:10
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  └ promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:530:36
```

Any ideas or suggestions?

jdm commented:
@Coder206 The main problem you're seeing is due to using the `test` directory (which doesn't exist) instead of `tests` (which does exist).

Coder206 commented on tests/wpt/mozilla/tests/mozilla/service-workers/service-worker-registration.html:
```diff
@@ -14,39 +14,42 @@
-              function() { var sw_reg = register_sw('sw.js', './mal%5fformed/sco%5Cpe/'); },
-              "Invalid URL Path");
+promise_test(function (p) {
+  return promise_rejects(p, new TypeError(), register_sw('sw.js', './mal%5fformed/sco%5Cpe/'));
```
@jdm Same for this one.

jdm commented on tests/wpt/mozilla/tests/mozilla/service-workers/service-worker-registration.html:
```diff
@@ -14,39 +14,42 @@
+  });
 }, "Test: Active Service Worker ScriptURL property");
 
+function bar() {
```
This function isn't used anywhere.

Coder206 replied:
@jdm Yes, sorry it was test code that escaped... Will remove it asap.

jdm reviewed (changes requested):
One teeny tiny change and then we can merge this!

Coder206 commented:
@jdm So I made the last changes requested and tested `mach build -d` and `./mach test-wpt tests/wpt/mozilla/tests/mozilla/service-workers/service-worker-registration.html`  both yielded success. Anything else to test or fix?

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit e6b8790 has been approved by `jdm`

<!-- @bors-servo r=jdm e6b879048fa7224885605cd695ca5b853799f55a -->

bors-servo commented:
:hourglass: Testing commit e6b8790 with merge 0d46c7c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/532)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/incremental_trailing_whitespace_a.html
      └   → /_mozilla/css/incremental_trailing_whitespace_a.html da54041086f2975f8e3c776d2283ad6609e6862a
    /_mozilla/css/incremental_trailing_whitespace_ref.html f5d0147c8bddbdc772a7e5e86c7c9e433fcd486b
    Testing da54041086f2975f8e3c776d2283ad6609e6862a == f5d0147c8bddbdc772a7e5e86c7c9e433fcd486b

jdm commented:
@bors-servo: retry
* #10473

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3807), [arm64](http://build.servo.org/builders/arm64/builds/3815), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3893), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/533), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3884), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3876), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/565), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/571), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1797) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/532)...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3807), [arm64](http://build.servo.org/builders/arm64/builds/3815), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3893), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/533), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/533), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3884), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3876), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/565), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/571), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1797)
